File: packages/llm/src/messaging/spam-guard.ts
import {
	FORBIDDEN_OPENERS,
	FORBIDDEN_PHRASES,
	FORBIDDEN_WORDS,
	MAX_EXCLAMATION_MARKS,
	MAX_QUESTION_MARKS,
	MAX_WORDS,
	MIN_QUALITY_SCORE,
} from "./config";
import type { ListingSignals } from "./models";
import type { LLMClient } from "./message-generator";

export interface ValidationResult {
	passed: boolean;
	score: number;
	rejectionReasons: string[];
}

export class SpamGuard {
	private llm: LLMClient | null;

	constructor(llmClient: LLMClient | null = null) {
		this.llm = llmClient;
	}

	async validate(
		message: string,
		signals: ListingSignals,
	): Promise<ValidationResult> {
		const reasons: string[] = [];

		// Layer 1: Rule-based checks
		reasons.push(...this.checkForbiddenWords(message));
		reasons.push(...this.checkForbiddenPhrases(message));
		reasons.push(...this.checkForbiddenOpeners(message));
		reasons.push(...this.checkStructure(message));
		reasons.push(...this.checkPersonalization(message, signals));
		reasons.push(...this.checkSelfFocus(message));

		if (reasons.length > 0) {
			return { passed: false, score: 0, rejectionReasons: reasons };
		}

		// Layer 2: LLM quality check (if client available)
		if (this.llm) {
			const score = await this.llmQualityCheck(message);
			if (score < MIN_QUALITY_SCORE) {
				return {
					passed: false,
					score,
					rejectionReasons: [
						`LLM-Qualitätsscore ${score}/10 -- unter Minimum von ${MIN_QUALITY_SCORE}`,
					],
				};
			}
			return { passed: true, score, rejectionReasons: [] };
		}

		return { passed: true, score: 0, rejectionReasons: [] };
	}

	private checkForbiddenWords(message: string): string[] {
		const reasons: string[] = [];
		const msgLower = message.toLowerCase();
		for (const word of FORBIDDEN_WORDS) {
			if (msgLower.includes(word.toLowerCase())) {
				reasons.push(`Verbotenes Wort: '${word}'`);
			}
		}
		return reasons;
	}

	private checkForbiddenPhrases(message: string): string[] {
		const reasons: string[] = [];
		const msgLower = message.toLowerCase();
		for (const phrase of FORBIDDEN_PHRASES) {
			if (msgLower.includes(phrase.toLowerCase())) {
				reasons.push(`Verbotene Phrase: '${phrase}'`);
			}
		}
		return reasons;
	}

	private checkForbiddenOpeners(message: string): string[] {
		const stripped = message.trim();
		for (const opener of FORBIDDEN_OPENERS) {
			if (stripped.startsWith(opener)) {
				return [`Verbotener Anfang: '${opener}...'`];
			}
		}
		return [];
	}

	private checkStructure(message: string): string[] {
		const reasons: string[] = [];
		const wordCount = message.split(/\s+/).length;

		if (wordCount > MAX_WORDS) {
			reasons.push(`Zu lang: ${wordCount} Wörter (max ${MAX_WORDS})`);
		}

		const exclCount = (message.match(/!/g) ?? []).length;
		if (exclCount > MAX_EXCLAMATION_MARKS) {
			reasons.push(
				`Zu viele Ausrufezeichen: ${exclCount} (max ${MAX_EXCLAMATION_MARKS})`,
			);
		}

		const questionCount = (message.match(/\?/g) ?? []).length;
		if (questionCount > MAX_QUESTION_MARKS) {
			reasons.push(
				`Zu viele Fragezeichen: ${questionCount} (max ${MAX_QUESTION_MARKS})`,
			);
		}

		if (questionCount === 0) {
			reasons.push(
				"Kein Fragezeichen -- Nachricht braucht mindestens eine Frage",
			);
		}

		// Message should contain a question but doesn't need to END with ?
		// because a greeting/sign-off can follow the question (Erstkontakt)

		if (/https?:\/\/|www\./i.test(message)) {
			reasons.push("Enthält URL -- nicht erlaubt");
		}

		if (message.includes("\u2014") || message.includes("\u2013")) {
			reasons.push("Em-dash/en-dash gefunden (AI-Tell)");
		}

		if (/\s--\s/.test(message)) {
			reasons.push(
				"Gedankenstrich (--) gefunden, keine Gedankenstriche benutzen",
			);
		}

		// Greeting endings are now ALLOWED for first contact (Erstkontakt)
		// Only block overly formal closings
		const tooFormalEndings = ["mit freundlichen grüßen", "hochachtungsvoll"];
		const msgLower = message.toLowerCase().trimEnd();
		const lastChars = msgLower.slice(-60);
		for (const ending of tooFormalEndings) {
			if (lastChars.includes(ending)) {
				reasons.push(`Zu formelle Grußformel: '${ending}'`);
			}
		}

		return reasons;
	}

	private checkPersonalization(
		message: string,
		signals: ListingSignals,
	): string[] {
		const msgLower = message.toLowerCase();

		// Check if any unique feature is mentioned
		for (const feature of signals.uniqueFeatures.slice(0, 5)) {
			const words = feature.split(/\s+/).filter((w) => w.length > 4);
			if (words.some((w) => msgLower.includes(w.toLowerCase()))) return [];
		}

		// Check if price is mentioned
		if (
			signals.price &&
			message.replace(/\./g, "").includes(String(signals.price))
		)
			return [];

		// Check if price in formatted form is mentioned
		if (signals.price) {
			const formatted = signals.price.toLocaleString("de-DE");
			if (message.includes(formatted)) return [];
		}

		// Check if city is mentioned
		if (signals.city && msgLower.includes(signals.city.toLowerCase()))
			return [];

		// Check if PLZ is mentioned
		if (signals.plz && message.includes(signals.plz)) return [];

		// Check if wohnfläche is mentioned
		if (
			signals.wohnflaeche &&
			message.includes(String(Math.round(signals.wohnflaeche)))
		)
			return [];

		// Check if grundstück is mentioned
		if (
			signals.grundstueck &&
			message.includes(String(Math.round(signals.grundstueck)))
		)
			return [];

		return [
			"Keine Personalisierung -- kein spezifisches Detail aus der Anzeige gefunden",
		];
	}

	private checkSelfFocus(message: string): string[] {
		const first50 = message.trim().slice(0, 50).toLowerCase();
		const selfWords = first50.match(
			/\b(ich|mein|mir|mich|meine|meinem|meinen|meiner)\b/g,
		);
		if (selfWords && selfWords.length >= 2) {
			return [
				"Zu viel Ich-Fokus am Anfang -- beginne mit der Immobilie, nicht mit dir",
			];
		}
		return [];
	}

	private async llmQualityCheck(message: string): Promise<number> {
		const systemPrompt = `\
Du bist ein privater Immobilienverkäufer auf Kleinanzeigen. Du bekommst regelmäßig \
Nachrichten von Maklern -- die meisten sind generische Copy-Paste-Templates die du ignorierst.

Bewerte diese Nachricht auf einer Skala von 1-10:
- 1-3: Generisches Makler-Template, offensichtlicher Sales-Pitch, würde ich ignorieren
- 4-5: Irgendein Makler, aber nichts Besonderes -- wahrscheinlich ignorieren
- 6-7: Klingt echt, hat was Spezifisches zu meiner Immobilie gesagt -- könnte antworten
- 8-10: Klingt wie ein normaler Mensch der zufällig Makler ist, hat mir was Nützliches \
gesagt oder mich neugierig gemacht -- würde antworten

Antworte NUR mit der Zahl (1-10), nichts weiter.`;

		try {
			const response = await this.llm!.generate(systemPrompt, message);
			const match = response.trim().match(/\b(\d+)\b/);
			if (match) {
				const score = Number.parseInt(match[1]);
				return Math.max(1, Math.min(10, score));
			}
			return 5;
		} catch {
			return 5;
		}
	}
}

File: dings/src/messaging/spam_guard.py
"""Validates generated messages against anti-spam rules and quality checks."""

from __future__ import annotations

import re
from dataclasses import dataclass

from .config import (
    FORBIDDEN_OPENERS,
    FORBIDDEN_PHRASES,
    FORBIDDEN_WORDS,
    MAX_EXCLAMATION_MARKS,
    MAX_QUESTION_MARKS,
    MAX_WORDS,
    MIN_QUALITY_SCORE,
)
from .models import ListingSignals


@dataclass
class ValidationResult:
    passed: bool
    score: int = 0  # 1-10 quality score from LLM
    rejection_reasons: list[str] | None = None

    @property
    def rejection_reason(self) -> str:
        if self.rejection_reasons:
            return "; ".join(self.rejection_reasons)
        return ""


class SpamGuard:
    """Two-layer validation: rule-based checks + LLM quality scoring."""

    def __init__(self, llm_client=None):
        """
        Args:
            llm_client: An object with a `generate(system_prompt, user_prompt) -> str`
                method. If None, LLM quality check is skipped.
        """
        self._llm = llm_client

    def validate(self, message: str, signals: ListingSignals) -> ValidationResult:
        """Run all validation checks on a generated message."""
        reasons: list[str] = []

        # Layer 1: Rule-based checks
        reasons.extend(self._check_forbidden_words(message))
        reasons.extend(self._check_forbidden_phrases(message))
        reasons.extend(self._check_forbidden_openers(message))
        reasons.extend(self._check_structure(message))
        reasons.extend(self._check_personalization(message, signals))
        reasons.extend(self._check_self_focus(message))

        if reasons:
            return ValidationResult(passed=False, score=0, rejection_reasons=reasons)

        # Layer 2: LLM quality check (if client available)
        if self._llm:
            score = self._llm_quality_check(message)
            if score < MIN_QUALITY_SCORE:
                return ValidationResult(
                    passed=False,
                    score=score,
                    rejection_reasons=[
                        f"LLM-Qualitätsscore {score}/10 — unter Minimum von {MIN_QUALITY_SCORE}"
                    ],
                )
            return ValidationResult(passed=True, score=score)

        return ValidationResult(passed=True, score=0)

    def _check_forbidden_words(self, message: str) -> list[str]:
        reasons = []
        msg_lower = message.lower()
        for word in FORBIDDEN_WORDS:
            if word.lower() in msg_lower:
                reasons.append(f"Verbotenes Wort: '{word}'")
        return reasons

    def _check_forbidden_phrases(self, message: str) -> list[str]:
        reasons = []
        msg_lower = message.lower()
        for phrase in FORBIDDEN_PHRASES:
            if phrase.lower() in msg_lower:
                reasons.append(f"Verbotene Phrase: '{phrase}'")
        return reasons

    def _check_forbidden_openers(self, message: str) -> list[str]:
        stripped = message.strip()
        for opener in FORBIDDEN_OPENERS:
            if stripped.startswith(opener):
                return [f"Verbotener Anfang: '{opener}...'"]
        return []

    def _check_structure(self, message: str) -> list[str]:
        reasons = []
        word_count = len(message.split())

        if word_count > MAX_WORDS:
            reasons.append(f"Zu lang: {word_count} Wörter (max {MAX_WORDS})")

        excl_count = message.count("!")
        if excl_count > MAX_EXCLAMATION_MARKS:
            reasons.append(
                f"Zu viele Ausrufezeichen: {excl_count} (max {MAX_EXCLAMATION_MARKS})"
            )

        question_count = message.count("?")
        if question_count > MAX_QUESTION_MARKS:
            reasons.append(
                f"Zu viele Fragezeichen: {question_count} (max {MAX_QUESTION_MARKS})"
            )

        if question_count == 0:
            reasons.append("Kein Fragezeichen — Nachricht braucht genau einen CTA")

        if not message.rstrip().endswith("?"):
            reasons.append("Nachricht muss mit einer Frage enden")

        # Check for URLs
        if re.search(r"https?://|www\.", message, re.IGNORECASE):
            reasons.append("Enthält URL — nicht erlaubt")

        # Check for greeting at end
        greeting_endings = [
            "viele grüße", "liebe grüße", "mit freundlichen grüßen",
            "mfg", "lg", "beste grüße", "herzliche grüße",
        ]
        msg_lower = message.lower().rstrip().rstrip("?").rstrip()
        for greeting in greeting_endings:
            if greeting in msg_lower[-50:]:
                reasons.append(f"Grußformel am Ende: '{greeting}'")

        return reasons

    def _check_personalization(self, message: str, signals: ListingSignals) -> list[str]:
        """Verify the message contains at least one specific detail from the listing."""
        msg_lower = message.lower()

        # Check if any unique feature is mentioned
        for feature in signals.unique_features[:5]:
            # Check individual words from the feature (at least 1 distinctive word)
            words = [w for w in feature.split() if len(w) > 4]
            if any(w.lower() in msg_lower for w in words):
                return []

        # Check if price is mentioned
        if signals.price and str(signals.price) in message.replace(".", ""):
            return []

        # Check if price in formatted form (385.000) is mentioned
        if signals.price:
            formatted = f"{signals.price:,}".replace(",", ".")
            if formatted in message:
                return []

        # Check if city is mentioned
        if signals.city and signals.city.lower() in msg_lower:
            return []

        # Check if PLZ is mentioned
        if signals.plz and signals.plz in message:
            return []

        # Check if wohnfläche is mentioned
        if signals.wohnflaeche and str(int(signals.wohnflaeche)) in message:
            return []

        # Check if grundstück is mentioned
        if signals.grundstueck and str(int(signals.grundstueck)) in message:
            return []

        return ["Keine Personalisierung — kein spezifisches Detail aus der Anzeige gefunden"]

    def _check_self_focus(self, message: str) -> list[str]:
        """Check if the message starts too self-focused."""
        first_50 = message.strip()[:50].lower()
        # Count first-person references in first 50 chars
        self_words = re.findall(r"\b(ich|mein|mir|mich|meine|meinem|meinen|meiner)\b", first_50)
        if len(self_words) >= 2:
            return ["Zu viel Ich-Fokus am Anfang — beginne mit der Immobilie, nicht mit dir"]
        return []

    def _llm_quality_check(self, message: str) -> int:
        """Score the message 1-10 from a seller's perspective using LLM."""
        system_prompt = """\
Du bist ein privater Immobilienverkäufer auf Kleinanzeigen. Du bekommst täglich \
30-50 Nachrichten, davon 80% Lowball-Angebote und Spam von Maklern.

Bewerte diese Nachricht auf einer Skala von 1-10:
- 1-3: Offensichtlicher Spam/Makler, würde ich ignorieren
- 4-5: Unklar, wahrscheinlich ignorieren
- 6-7: Interessant, könnte antworten
- 8-10: Würde definitiv antworten, fühlt sich echt an

Antworte NUR mit der Zahl (1-10), nichts weiter."""

        try:
            response = self._llm.generate(system_prompt, message)
            # Extract the first number from the response
            match = re.search(r"\b(\d+)\b", response.strip())
            if match:
                score = int(match.group(1))
                return max(1, min(10, score))
            return 5  # Default if parsing fails
        except Exception:
            return 5  # Default on error — don't block message
